objectstore-server/src/http.rs

```diff
@@ -20,56 +20,120 @@ use tokio::net::{TcpListener, TcpSocket};
+        // builder order defines layers added first will be called first. This means:
+        //  - Requests go from top to bottom
+        //  - Responses go from bottom to top
+        let middleware = ServiceBuilder::new()
+            .layer(axum::middleware::from_fn(emit_request_metrics))
+            .layer(in_flight_layer)
+            .layer(CatchPanicLayer::custom(handle_panic))
+            .layer(SetResponseHeaderLayer::overriding(
+                header::SERVER,
+                HeaderValue::from_static(SERVER),
+            ))
+            .layer(NewSentryLayer::new_from_top())
+            .layer(SentryHttpLayer::new().enable_transaction())
+            .layer(
+                TraceLayer::new_for_http()
+                    .make_span_with(make_http_span)
+                    .on_failure(DefaultOnFailure::new().level(Level::DEBUG)),
+            );
+
+        let service_routes = Router::new().route("/", put(put_object_nokey)).route(
+            "/{*key}",
+            put(put_object).get(get_object).delete(delete_object),
         );
 
-    let service_routes = Router::new().route("/", put(put_object_nokey)).route(
-        "/{*key}",
-        put(put_object).get(get_object).delete(delete_object),
-    );
+        let router = Router::new()
+            .nest("/v1/", service_routes)
+            .route("/health", get(health))
+            .layer(middleware)
+            .with_state(state);
 
-    Router::new()
-        .nest("/v1/", service_routes)
-        .route("/health", get(health))
-        .layer(middleware)
-        .with_state(state)
+        App {
+            router,
+            in_flight_requests,
+            graceful_shutdown: false,
+        }
+    }
+
+    /// Enables or disables graceful shutdown for the server.
+    ///
+    /// By default, graceful shutdown is disabled.
+    pub fn graceful_shutdown(mut self, enable: bool) -> Self {
+        self.graceful_shutdown = enable;
+        self
+    }
+
+    /// Runs the web server until graceful shutdown is triggered.
+    ///
+    /// This function creates a future that runs the server. The future must be spawned or awaited for
+    /// the server to continue running.
+    pub async fn serve(self, listener: TcpListener) -> Result<()> {
+        let Self {
+            router,
+            in_flight_requests,
+            graceful_shutdown,
+        } = self;
+
+        let service =
+            ServiceExt::<Request>::into_make_service_with_connect_info::<SocketAddr>(router);
+
+        let server = async move {
+            if graceful_shutdown {
+                let guard = elegant_departure::get_shutdown_guard().shutdown_on_drop();
+                axum::serve(listener, service)
+                    .with_graceful_shutdown(guard.wait_owned())
+                    .await
+            } else {
+                axum::serve(listener, service).await
+            }
+        };
```

Comment:
This became much more verbose than I would've liked, but I cannot see another way given the types with and without graceful shutdown are different.